File: threedollar-api-user-service/src/e2eTest/java/com/depromeet/threedollar/api/userservice/controller/auth/AuthControllerTest.java
package com.depromeet.threedollar.api.userservice.controller.auth;

import com.depromeet.threedollar.api.core.common.dto.ApiResponse;
import com.depromeet.threedollar.api.userservice.SetupUserControllerTest;
import com.depromeet.threedollar.api.core.listener.commonservice.device.DeviceEventListener;
import com.depromeet.threedollar.api.userservice.service.auth.dto.request.LoginRequest;
import com.depromeet.threedollar.api.userservice.service.auth.dto.request.SignUpRequest;
import com.depromeet.threedollar.api.userservice.service.auth.dto.response.LoginResponse;
import com.depromeet.threedollar.domain.rds.domain.userservice.user.User;
import com.depromeet.threedollar.domain.rds.domain.userservice.user.UserFixture;
import com.depromeet.threedollar.domain.rds.domain.userservice.user.UserSocialType;
import com.depromeet.threedollar.domain.rds.event.userservice.user.UserLogOutedEvent;
import com.depromeet.threedollar.domain.rds.event.userservice.user.UserSignOutedEvent;
import com.depromeet.threedollar.infrastructure.external.client.apple.AppleTokenDecoder;
import com.depromeet.threedollar.infrastructure.external.client.google.GoogleAuthApiClient;
import com.depromeet.threedollar.infrastructure.external.client.google.dto.response.GoogleProfileInfoResponse;
import com.depromeet.threedollar.infrastructure.external.client.kakao.KaKaoAuthApiClient;
import com.depromeet.threedollar.infrastructure.external.client.kakao.dto.response.KaKaoProfileResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.mock.mockito.MockBean;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class AuthControllerTest extends SetupUserControllerTest {

    private static final String KAKAO_SOCIAL_ID = "kakao-social-id";
    private static final String GOOGLE_SOCIAL_ID = "google-social-id";
    private static final String APPLE_SOCIAL_ID = "apple-social-id";

    private AuthMockApiCaller authMockApiCaller;

    @MockBean
    private KaKaoAuthApiClient kaKaoAuthApiClient;

    @MockBean
    private AppleTokenDecoder appleTokenDecoder;

    @MockBean
    private GoogleAuthApiClient googleAuthApiClient;

    @MockBean
    private DeviceEventListener deviceEventListener;

    @BeforeEach
    void setUp() {
        authMockApiCaller = new AuthMockApiCaller(mockMvc, objectMapper);
    }

    @BeforeEach
    void mockAuthApiClient() {
        when(kaKaoAuthApiClient.getProfileInfo(anyString())).thenReturn(KaKaoProfileResponse.testInstance(KAKAO_SOCIAL_ID));
        when(appleTokenDecoder.getSocialIdFromIdToken(anyString())).thenReturn(APPLE_SOCIAL_ID);
        when(googleAuthApiClient.getProfileInfo(anyString())).thenReturn(GoogleProfileInfoResponse.testInstance(GOOGLE_SOCIAL_ID));
    }

    @Nested
    class KaKaoAuthApiTest {

        @Test
        void 카카오_회원가입_요청이_성공하면_인증_토큰이_반환된다() throws Exception {
            // given
            SignUpRequest request = SignUpRequest.testBuilder()
                .token("kakao-token")
                .name("가슴속 삼천원")
                .socialType(UserSocialType.KAKAO)
                .build();

            // when
            ApiResponse<LoginResponse> response = authMockApiCaller.signUp(request, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData().getToken()).isNotBlank()
            );
        }

        @Test
        void 카카오_로그인_요청이_성공하면_인증_토큰이_반환된다() throws Exception {
            // given
            User user = UserFixture.create(KAKAO_SOCIAL_ID, UserSocialType.KAKAO, "카카오 계정");
            userRepository.save(user);

            LoginRequest request = LoginRequest.testBuilder()
                .token("kakao-token")
                .socialType(UserSocialType.KAKAO)
                .build();

            // when
            ApiResponse<LoginResponse> response = authMockApiCaller.login(request, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData().getToken()).isNotBlank()
            );
        }

    }

    @Nested
    class AppleAuthApiTest {

        @Test
        void 애플_회원가입_요청이_성공하면_인증_토큰이_반환된다() throws Exception {
            // given
            SignUpRequest request = SignUpRequest.testBuilder()
                .token("apple-token")
                .name("가슴속 삼천원")
                .socialType(UserSocialType.APPLE)
                .build();

            // when
            ApiResponse<LoginResponse> response = authMockApiCaller.signUp(request, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData().getToken()).isNotBlank()
            );
        }

        @Test
        void 애플_로그인_요청이_성공하면_인증_토큰이_반환된다() throws Exception {
            // given
            User user = UserFixture.create(APPLE_SOCIAL_ID, UserSocialType.APPLE, "애플 계정");
            userRepository.save(user);

            LoginRequest request = LoginRequest.testBuilder()
                .token("apple-token")
                .socialType(UserSocialType.APPLE)
                .build();

            // when
            ApiResponse<LoginResponse> response = authMockApiCaller.login(request, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData().getToken()).isNotBlank()
            );
        }

    }

    @Nested
    class GoogleAuthApiTest {

        @Test
        void 구글_회원가입_요청이_성공하면_인증_토큰이_반환된다() throws Exception {
            // given
            SignUpRequest request = SignUpRequest.testBuilder()
                .token("google-token")
                .name("가슴속 삼천원")
                .socialType(UserSocialType.GOOGLE)
                .build();

            // when
            ApiResponse<LoginResponse> response = authMockApiCaller.signUp(request, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData().getToken()).isNotBlank()
            );
        }

        @Test
        void 구글_로그인_요청이_성공하면_인증_토큰이_반환된다() throws Exception {
            // given
            User user = UserFixture.create(GOOGLE_SOCIAL_ID, UserSocialType.GOOGLE, "구글 계정");
            userRepository.save(user);

            LoginRequest request = LoginRequest.testBuilder()
                .token("google-access-token")
                .socialType(UserSocialType.GOOGLE)
                .build();

            // when
            ApiResponse<LoginResponse> response = authMockApiCaller.login(request, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData().getToken()).isNotBlank()
            );
        }

    }

    @DisplayName("DELETE /api/v2/signout")
    @Nested
    class SignOutApiTest {

        @Test
        void 회원탈퇴_요청이_성공하면_200_OK() throws Exception {
            // when
            ApiResponse<String> response = authMockApiCaller.signOut(token, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData()).isEqualTo(ApiResponse.OK.getData())
            );
        }

        @Test
        void 로그아웃시_디바이스_삭제_이벤트가_발송된다() throws Exception {
            // when
            authMockApiCaller.signOut(token, 200);

            // then
            verify(deviceEventListener, times(1)).deleteUserDevice(any(UserSignOutedEvent.class));
        }

    }

    @DisplayName("DELETE /api/v2/logout")
    @Nested
    class LogoutApiTest {

        @Test
        void 로그아웃_성공시_200_OK() throws Exception {
            // when
            ApiResponse<String> response = authMockApiCaller.logout(token, 200);

            // then
            assertAll(
                () -> assertThat(response.getResultCode()).isEmpty(),
                () -> assertThat(response.getMessage()).isEmpty(),
                () -> assertThat(response.getData()).isEqualTo(ApiResponse.OK.getData())
            );
        }

        @Test
        void 로그아웃시_디바이스_삭제_이벤트가_발송된다() throws Exception {
            // when
            authMockApiCaller.logout(token, 200);

            // then
            verify(deviceEventListener, times(1)).deleteUserDevice(any(UserLogOutedEvent.class));
        }

    }

}

File: threedollar-api-boss-service/src/e2e-test/kotlin/com/depromeet/threedollar/api/bossservice/controller/auth/AuthControllerTest.kt
package com.depromeet.threedollar.api.bossservice.controller.auth

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.delete
import org.springframework.test.web.servlet.post
import com.depromeet.threedollar.api.bossservice.SetupBossAccountControllerTest
import com.depromeet.threedollar.api.bossservice.service.auth.dto.request.LoginRequest
import com.depromeet.threedollar.api.bossservice.service.auth.dto.request.SignupRequest
import com.depromeet.threedollar.common.exception.type.ErrorCode
import com.depromeet.threedollar.domain.mongo.domain.bossservice.account.BossAccountCreator
import com.depromeet.threedollar.domain.mongo.domain.bossservice.account.BossAccountSocialType
import com.depromeet.threedollar.domain.mongo.domain.bossservice.registration.BossRegistrationRepository
import com.depromeet.threedollar.domain.mongo.domain.bossservice.registration.RegistrationCreator
import com.depromeet.threedollar.external.client.apple.AppleTokenDecoder
import com.depromeet.threedollar.external.client.google.GoogleAuthApiClient
import com.depromeet.threedollar.external.client.google.dto.response.GoogleProfileInfoResponse
import com.depromeet.threedollar.external.client.kakao.KaKaoAuthApiClient
import com.depromeet.threedollar.external.client.kakao.dto.response.KaKaoProfileResponse
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every

internal class AuthControllerTest(
    private val bossRegistrationRepository: BossRegistrationRepository,
) : SetupBossAccountControllerTest() {

    @MockkBean
    private lateinit var kaKaoAuthApiClient: KaKaoAuthApiClient

    @MockkBean
    private lateinit var googleAuthApiClient: GoogleAuthApiClient

    @MockkBean
    private lateinit var appleTokenDecoder: AppleTokenDecoder

    @AfterEach
    fun cleanUp() {
        super.cleanup()
        bossRegistrationRepository.deleteAll()
    }

    @DisplayName("POST /api/v2/signup")
    @Nested
    inner class SignupApiTest {

        @Test
        fun 카카오_회원가입_요청이_성공하면_토큰이_반환된다() {
            // given
            every { kaKaoAuthApiClient.getProfileInfo(any()) } returns KaKaoProfileResponse.testInstance("kakao-socialId")

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.KAKAO,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 카카오_회원가입시_이미_존재하는_사장님_계정인경우_409_에러가_발생한다() {
            // given
            val boss = BossAccountCreator.create("google-social-id", BossAccountSocialType.KAKAO, "카카오 계정")
            bossAccountRepository.save(boss)

            every { kaKaoAuthApiClient.getProfileInfo(any()) } returns KaKaoProfileResponse.testInstance(boss.socialInfo.socialId)

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.KAKAO,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isConflict() }
                    jsonPath("$.resultCode") { ErrorCode.CONFLICT_BOSS_ACCOUNT.code }
                    jsonPath("$.resultCode") { ErrorCode.CONFLICT_BOSS_ACCOUNT.message }
                }
        }

        @Test
        fun 카카오_회원가입시_가입_승인_대기중인경우_403_에러가_발생한다() {
            // given
            val registration = RegistrationCreator.create("google-social-id", BossAccountSocialType.KAKAO)
            bossRegistrationRepository.save(registration)

            every { kaKaoAuthApiClient.getProfileInfo(any()) } returns KaKaoProfileResponse.testInstance(registration.boss.socialInfo.socialId)

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.KAKAO,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isForbidden() }
                    jsonPath("$.resultCode") { ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT.code }
                    jsonPath("$.resultCode") { ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT.message }
                }
        }

        @Test
        fun 애플_회원가입_요청이_성공하면_토큰이_반환된다() {
            // given
            every { appleTokenDecoder.getSocialIdFromIdToken(any()) } returns "apple-social-id"

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.APPLE,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 애플_회원가입시_이미_존재하는_사장님_계정인경우_409_에러가_발생한다() {
            // given
            val boss = BossAccountCreator.create("apple-social-id", BossAccountSocialType.APPLE, "애플 계정")
            bossAccountRepository.save(boss)

            every { appleTokenDecoder.getSocialIdFromIdToken(any()) } returns boss.socialInfo.socialId

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.APPLE,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isConflict() }
                    jsonPath("$.resultCode") { ErrorCode.CONFLICT_BOSS_ACCOUNT.code }
                    jsonPath("$.resultCode") { ErrorCode.CONFLICT_BOSS_ACCOUNT.message }
                }
        }

        @Test
        fun 애플_회원가입시_가입_승인_대기중인경우_403_에러가_발생한다() {
            // given
            val registration = RegistrationCreator.create("apple-social-id", BossAccountSocialType.APPLE)
            bossRegistrationRepository.save(registration)

            every { appleTokenDecoder.getSocialIdFromIdToken(any()) } returns registration.boss.socialInfo.socialId

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.APPLE,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isForbidden() }
                    jsonPath("$.resultCode") { ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT.code }
                    jsonPath("$.resultCode") { ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT.message }
                }
        }

        @Test
        fun 구글_회원가입_요청이_성공하면_토큰이_반환된다() {
            // given
            every { googleAuthApiClient.getProfileInfo(any()) } returns GoogleProfileInfoResponse.testInstance("google-social-id")

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.GOOGLE,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 구글_회원가입시_이미_존재하는_사장님_계정인경우_409_에러가_발생한다() {
            // given
            val boss = BossAccountCreator.create("google-social-id", BossAccountSocialType.GOOGLE, "구글 계정")
            bossAccountRepository.save(boss)

            every { googleAuthApiClient.getProfileInfo(any()) } returns GoogleProfileInfoResponse.testInstance(boss.socialInfo.socialId)

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.GOOGLE,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isConflict() }
                    jsonPath("$.resultCode") { ErrorCode.CONFLICT_BOSS_ACCOUNT.code }
                    jsonPath("$.resultCode") { ErrorCode.CONFLICT_BOSS_ACCOUNT.message }
                }
        }

        @Test
        fun 구글_회원가입시_가입_승인_대기중인경우_403_에러가_발생한다() {
            // given
            val registration = RegistrationCreator.create("google-social-id", BossAccountSocialType.GOOGLE)
            bossRegistrationRepository.save(registration)

            every { googleAuthApiClient.getProfileInfo(any()) } returns GoogleProfileInfoResponse.testInstance(registration.boss.socialInfo.socialId)

            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.GOOGLE,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isForbidden() }
                    jsonPath("$.resultCode") { ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT.code }
                    jsonPath("$.resultCode") { ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT.message }
                }
        }

        @Test
        fun 네이버_회원가입_요청시_ServiceUnavailable_에러가_발생한다() {
            // given
            val request = SignupRequest(
                token = "token",
                socialType = BossAccountSocialType.NAVER,
                bossName = "bossName",
                businessNumber = "000-00-00000",
                storeName = "가게 이름",
                storeCategoriesIds = setOf(),
                contactsNumber = "010-1234-1234",
                certificationPhotoUrl = "https://photo.png"
            )

            // when & then
            mockMvc.post("/v1/auth/signup") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isServiceUnavailable() }
                    jsonPath("$.resultCode") { value(ErrorCode.SERVICE_UNAVAILABLE.code) }
                    jsonPath("$.message") { value(ErrorCode.SERVICE_UNAVAILABLE.message) }
                }
        }

    }

    @DisplayName("POST /api/v2/login")
    @Nested
    inner class LoginApiTest {

        @Test
        fun 카카오_로그인_요청이_성공하면_토큰이_반환된다() {
            // given
            val boss = BossAccountCreator.create("google-social-id", BossAccountSocialType.KAKAO, "카카오 계정")
            bossAccountRepository.save(boss)

            every { kaKaoAuthApiClient.getProfileInfo(any()) } returns KaKaoProfileResponse.testInstance(boss.socialInfo.socialId)

            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.KAKAO)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.bossId") { value(boss.id) }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 카카오_로그인시_가입승인_대기중인경우_토큰이_반환된다() {
            // given
            val registration = RegistrationCreator.create("google-social-id", BossAccountSocialType.KAKAO)
            bossRegistrationRepository.save(registration)

            every { kaKaoAuthApiClient.getProfileInfo(any()) } returns KaKaoProfileResponse.testInstance(registration.boss.socialInfo.socialId)

            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.KAKAO)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.bossId") { value(registration.id) }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 애플_로그인_요청이_성공하면_토큰이_반환된다() {
            // given
            val boss = BossAccountCreator.create("google-social-id", BossAccountSocialType.APPLE, "애플 계정")
            bossAccountRepository.save(boss)

            every { appleTokenDecoder.getSocialIdFromIdToken(any()) } returns boss.socialInfo.socialId

            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.APPLE)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.bossId") { value(boss.id) }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 애플_로그인시_가입승인_대기중인경우_토큰이_반환된다() {
            // given
            val registration = RegistrationCreator.create("google-social-id", BossAccountSocialType.APPLE)
            bossRegistrationRepository.save(registration)

            every { appleTokenDecoder.getSocialIdFromIdToken(any()) } returns registration.boss.socialInfo.socialId

            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.APPLE)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.bossId") { value(registration.id) }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 구글_로그인_요청이_성공하면_토큰이_반환된다() {
            // given
            val boss = BossAccountCreator.create("google-social-id", BossAccountSocialType.GOOGLE, "구글 계정")
            bossAccountRepository.save(boss)

            every { googleAuthApiClient.getProfileInfo(any()) } returns GoogleProfileInfoResponse.testInstance(boss.socialInfo.socialId)

            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.GOOGLE)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.bossId") { value(boss.id) }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 구글_로그인시_가입승인_대기중인경우_토큰이_반환된다() {
            // given
            val registration = RegistrationCreator.create("google-social-id", BossAccountSocialType.GOOGLE)
            bossRegistrationRepository.save(registration)

            every { googleAuthApiClient.getProfileInfo(any()) } returns GoogleProfileInfoResponse.testInstance(registration.boss.socialInfo.socialId)

            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.GOOGLE)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.data.bossId") { value(registration.id) }
                    jsonPath("$.data.token") { isNotEmpty() }
                }
        }

        @Test
        fun 네이버_로그인_요청시_ServiceUnavailable_에러가_발생한다() {
            // given
            val request = LoginRequest(token = "token", socialType = BossAccountSocialType.NAVER)

            // when & then
            mockMvc.post("/v1/auth/login") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isServiceUnavailable() }
                    jsonPath("$.resultCode") { value(ErrorCode.SERVICE_UNAVAILABLE.code) }
                    jsonPath("$.message") { value(ErrorCode.SERVICE_UNAVAILABLE.message) }
                }
        }

    }

    @DisplayName("POST /boss/v1/auth/logout")
    @Nested
    inner class LogoutApiTest {

        @Test
        fun 로그아웃_성공시_200_OK() {
            // when & then
            mockMvc.post("/v1/auth/logout")
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                }
        }

    }

    @DisplayName("DELETE /api/v2/signout")
    @Nested
    inner class SignOutApiTest {

        @Test
        fun 회원탈퇴_요청이_성공하면_200_OK() {
            // when & then
            mockMvc.delete("/v1/auth/signout") {
                header(HttpHeaders.AUTHORIZATION, "Bearer $token")
            }
                .andDo {
                    print()
                }
                .andExpect {
                    status { isOk() }
                }
        }

    }

}
